### PR TITLE
[test plan] 4-byte AS Translation

### PR DESCRIPTION
**Pull request by Azarack**
### Description of PR

Test plan and automation to verify 4-byte to 2-byte ASN translation feature functions as expected.

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [X] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### Any platform specific information?
vsonic neighbors 

#### Supported testbed topology if it's a new test case?
T2 topology

### CODE REVIEW

**Comment by Azarack** [comment]
@wenyiz2021 can you review this PR?

**Comment by hasan-brcm** [comment]
@kishorekunal01 review

File: tests/bgp/test_bgp_4-byte_as_trans.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    -c "neighbor {} timers connect 10" \
+    -c "neighbor {} remote-as {}" \
+    -c "neighbor {} peer-group {}" \
+    -c "neighbor {} description {}" \
+    -c "neighbor {} timers 3 10" \
+    -c "neighbor {} timers connect 10" \
+    -c "address-family ipv4 unicast" \
+    -c "network {}" \
+    -c "neighbor {} soft-reconfiguration inbound" \
+    -c "neighbor {} route-map FROM_BGP_PEER_V4 in" \
```
Comment (kishorekunal01) [inline tests/bgp/test_bgp_4-byte_as_trans.py:168]: route-map define is missing in the file, what exactly the route-map contain, similar other route map in the files as well.

Comment (Azarack) [reply tests/bgp/test_bgp_4-byte_as_trans.py:168]: This is a pre-configured route map that is part of the T2 BGP configuration.

File: tests/bgp/test_bgp_4-byte_as_trans.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    bgp_facts = duthost.bgp_facts(instance_id=asic_index)['ansible_facts']
+    for k, v in bgp_facts['bgp_neighbors'].items():
+        if v['description'].lower() not in skip_hosts:
+            logger.debug(v['description'])
+            assert v['state'] == 'established'
+
+
+def test_4_byte_asn_translation(setup):
+    # copy existing BGP config to a new 4-byte ASN.  Use existing route-maps for consistancy.
+    cmd = 'vtysh{} \
```
Comment (wenyiz2021) [inline tests/bgp/test_bgp_4-byte_as_trans.py:140]: there is a space needed after `vtysh` : cli_options = "-n " + duthost.get_namespace_from_asic_id(asic_index)

File: tests/bgp/test_bgp_4-byte_as_trans.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+             setup['neigh2_ip_v4'], setup['neigh2_asn'], setup['neigh2_ip_v4'], setup['peer_group_v4'],
+             setup['neigh2_ip_v4'], setup['neigh2'], setup['neigh2_ip_v4'], setup['neigh2_ip_v4'],
+             setup['neigh2_ip_v6'], setup['neigh2_asn'], setup['neigh2_ip_v6'], setup['peer_group_v6'],
+             setup['neigh2_ip_v6'], setup['neigh2'], setup['neigh2_ip_v6'], setup['neigh2_ip_v6'],
+             setup['dut_ipv4_network'], setup['peer_group_v4'], setup['peer_group_v4'], setup['peer_group_v4'],
+             setup['neigh_ip_v4'], setup['neigh2_ip_v4'], setup['dut_ipv6_network'], setup['peer_group_v6'],
+             setup['peer_group_v6'], setup['peer_group_v6'], setup['neigh_ip_v6'], setup['neigh2_ip_v6'])
+    logger.debug(setup['duthost'].shell(cmd, module_ignore_errors=True))
+
+    cmd = 'vtysh \
```
Comment (wenyiz2021) [inline tests/bgp/test_bgp_4-byte_as_trans.py:203]: can you add the `cli_options = "-n " + duthost.get_namespace_from_asic_id(asic_index)` here also? T2 neighbor could be multi-asic

File: tests/bgp/test_bgp_4-byte_as_trans.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+             setup['dut_ip_v6'], setup['dut_ip_v6'], setup['neigh_ipv4_network'], setup['peer_group_v4'],
+             setup['peer_group_v4'], setup['peer_group_v4'], setup['dut_ip_v4'], setup['neigh_ipv6_network'],
+             setup['peer_group_v6'], setup['peer_group_v6'], setup['peer_group_v6'], setup['dut_ip_v6'])
+
+    logger.debug(setup['neighhost'].shell(cmd, module_ignore_errors=True))
+
+    logger.debug("DUT BGP Config: {}".format(setup['duthost'].shell("show run bgp")['stdout']))
+    logger.debug("Neighbor BGP Config: {}".format(setup['neighhost'].shell("show run bgp")['stdout']))
+
+    time.sleep(bgp_sleep)
```
Comment (wenyiz2021) [inline tests/bgp/test_bgp_4-byte_as_trans.py:253]: could you elaborate on what the previous cmd is doing?
the sleep time could be different among T2s

File: tests/bgp/test_bgp_4-byte_as_trans.py (hunk 2 above)
Comment (Azarack) [reply tests/bgp/test_bgp_4-byte_as_trans.py:140]: Yes, sorry I missed this.

File: tests/bgp/test_bgp_4-byte_as_trans.py (hunk 4 above)
Comment (Azarack) [reply tests/bgp/test_bgp_4-byte_as_trans.py:253]: The previous command configures a 4-byte ASN on the neighbor device.  
I'm not sure what you mean by the sleep time being different as this is a variable set in the script.

**Comment by Azarack** [comment]
@wenyiz2021 Hello, can you please look at the updates I have made as per your comments?

File: tests/bgp/test_bgp_4-byte_as_trans.py (hunk 4 above)
Comment (wenyiz2021) [reply tests/bgp/test_bgp_4-byte_as_trans.py:259]: I mean how did you decide bgp_sleep?

Comment (Azarack) [reply tests/bgp/test_bgp_4-byte_as_trans.py:259]: It is the standard amount of time we wait for our tests to allow BGP to converge.  Then after running the test case if there is an issue it is extended.

Comment (wenyiz2021) [reply tests/bgp/test_bgp_4-byte_as_trans.py:259]: LGTM if this sleep time of 120 works well on T2

**Review by wenyiz2021 — APPROVED** [review]

**Comment by wenyiz2021** [comment]
LGTM

@kishorekunal01 can you check if your question is resolved

**Review by kishorekunal01 — APPROVED** [review]

**Comment by kishorekunal01** [comment]
> LGTM
> 
> @kishorekunal01 can you check if your question is resolved

Yes, it is resolved

**Comment by Azarack** [comment]
@wenyiz2021 Can you merge this or does someone else need to approve?

File: docs/testplan/BGP-4-Byte_AS_Translation.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
+- [Overview](#overview)
+    - [Scope](#scope)
+    - [Testbed](#testbed)
+- [Setup configuration](#setup-configuration)
+- [Test cases](#test-cases)
+
+# Test name
+
+BGP Session Flaps
```
Comment (prsunny) [inline docs/testplan/BGP-4-Byte_AS_Translation.md:9]: looks like the test name and overview is not aligned with test description. Can you check?

File: docs/testplan/BGP-4-Byte_AS_Translation.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+## Overview
+
+The goal of this test is to verify that the CPU and memory do not spike during cycles of BGP sessions flapping.
```
Comment (prsunny) [inline docs/testplan/BGP-4-Byte_AS_Translation.md:13]: Is this really meant to be a test for session flap and memory check?

File: docs/testplan/BGP-4-Byte_AS_Translation.md (hunk 5 above)
Comment (Azarack) [reply docs/testplan/BGP-4-Byte_AS_Translation.md]: @prsunny You are correct that this was a mistake.  I have updated it to the proper name.

File: docs/testplan/BGP-4-Byte_AS_Translation.md (hunk 6 above)
Comment (Azarack) [reply docs/testplan/BGP-4-Byte_AS_Translation.md]: @prsunny You are correct that this was a mistake.  I have updated it to the proper overview.

**Comment by Azarack** [comment]
@BYGX-wcr Please review this PR.

**Comment by wenyiz2021** [comment]
@Azarack I don't have permission to merge, if @prsunny approves he can merge

**Comment by Azarack** [comment]
> @Azarack I don't have permission to merge, if @prsunny approves he can merge

@prsunny Are you able to merge this PR now that it is approved?

**Review by prsunny — APPROVED** [review]

**Comment by prsunny** [comment]
@wangxin , would you please merge?

File: tests/bgp/test_bgp_4-byte_as_trans.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        neigh_cli_options = " -n " + neigh1.get_namespace_from_asic_id(asic_index)
+    else:
+        neigh_cli_options = ''
+
+    dut_ip_v4 = tbinfo['topo']['properties']['configuration'][neigh1]['bgp']['peers'][dut_asn][0]
+    dut_ip_v6 = tbinfo['topo']['properties']['configuration'][neigh1]['bgp']['peers'][dut_asn][1]
+
+    dut_ip_bgp_sum = duthost.shell('show ip bgp summary')['stdout']
+    neigh_ip_bgp_sum = nbrhosts[neigh1]["host"].shell('show ip bgp summary')['stdout']
+    neigh2_ip_bgp_sum = nbrhosts[neigh1]["host"].shell('show ip bgp summary')['stdout']
```
Comment (wenyiz2021) [inline tests/bgp/test_bgp_4-byte_as_trans.py:82]: needs to be nbrhosts[neigh2]?

**Pull request merged**